Readme.md
### svgPanZoom - [htmlwidget](http://htmlwidgets.org) to Pan / Zoom R graphics

[![Build Status](https://travis-ci.org/timelyportfolio/svgPanZoom.png?branch=master)](https://travis-ci.org/timelyportfolio/svgPanZoom)

`svgPanZoom` is a [`htmlwidgets`](http://htmlwidgets.org) wrapper for [`svg-pan-zoom.js`](https://github.com/ariutta/svg-pan-zoom).  `svgPanZoom` gives `R` users an easy way to add panning and zooming to any `R` graphics (base, ggplot2, lattice, and lots more).  It prioritizes ease and simplicity, but does offer some ability to tailor the experience and enhance the interactivity.

### Install It
For now (`svgPanZoom` not on CRAN) to get started, you will need to use `devtools::install_github` as shown below.  If you do not have `devtools`, please install with `install.packages("devtools")`.

```
devtools::install_github("timelyportfolio/svgPanZoom")
```

### Use It
As stated in the introduction `svgPanZoom` works with almost all `R` graphics types.  For `base` graphics, we'll need the `SVGAnnotation` package.

```
library(svgPanZoom) # see install step above
library(SVGAnnotation)

svgPanZoom(
  svgPlot(
    plot(1:10)
  )
)
```

### Use It in Shiny

There are lots more examples below, but real quickly here is how we can use it in Shiny.

```R
library(shiny)
library(SVGAnnotation)
library(svgPanZoom)
library(ggplot2)

ui <- shinyUI(bootstrapPage(
  
  svgPanZoomOutput(outputId = "main_plot")
  
))

server = shinyServer(function(input, output) {
  output$main_plot <- renderSvgPanZoom({
    p <- ggplot() + geom_point(data=data.frame(faithful),aes(x=eruptions,y=waiting)) + stat_density2d(data=data.frame(faithful),aes(x=eruptions,y=waiting, alpha =..level..),geom="polygon") + scale_alpha_continuous(range=c(0.05,0.2))
    svgPanZoom(p, controlIconsEnabled = T)
  })
})
  
runApp(list(ui=ui,server=server))
```

### Use It With Grid and More

Although `SVGAnnotation` works with `grid` graphics, such as `ggplot2` and `lattice`, we will **need to change the default** `addInfo` to `addInfo = F`.  Before I show an example though, I **highly recommend** using `gridSVG` for `ggplot2` and `lattice`.  For some good reasons, please see [this](http://stattech.wordpress.fos.auckland.ac.nz/2013-4-generating-structured-and-labelled-svg/) from Paul Murrell and Simon Potter.  If you are making big graphics--think maps, multiple graphs, etc.--for **speed stick with `svgPlot`**. Here is a simple example using `ggplot2` with `SVGAnnotation` and `svgPlot`.

```
library(svgPanZoom)
library(SVGAnnotation)
library(ggplot2)

svgPanZoom(
  svgPlot(
    #will put on separate line but also need show
    show(
      ggplot(data.frame(x=1:10,y=1:10),aes(x=x,y=y)) + geom_line()

    )
    # this is critical with svgPlot and grid graphics
    , addInfo = F
  )
)
```

Now let's do that same plot with `gridSVG`.

```
svgPanZoom(
  ggplot(data.frame(x=1:10,y=1:10),aes(x=x,y=y)) + geom_line()
)
```

You might notice right off that sizing is better handled, but more importantly, the resulting `SVG` is much better structured `XML`.  That small time lag will really start to hurt if you are using `gridSVG` with large or complicated graphics.  So if speed is important, ignore the better structure from `gridSVG` and stick with `svgPlot`.

As promised, `lattice` (yes, I still use it and like it) works just as nicely.

```
library(svgPanZoom)
library(SVGAnnotation)
library(lattice)
 
# with gridSVG
svgPanZoom(
  xyplot( y~x, data.frame(x=1:10,y=1:10), type = "b" )
)

# with svgPlot
svgPanZoom(
  svgPlot(
    show(xyplot( y~x, data.frame(x=1:10,y=1:10), type = "b" ))
    , height = 6
    , width = 10
  )
)
```

If you are not impressed yet, then maybe the graphics were not compelling enough.  Let's add `svgPanZoom` to some more complicated visualizations.

```
library(choroplethr)
# from chorplethr documetation
data(df_pop_state)
m = state_choropleth(
  df_pop_state
  , title="US 2012 State Population Estimates"
  , legend="Population"
)
# take a peek
m
# would be so much more fun with pan and zoom
svgPanZoom( m )
# if your map is big and hairy do this instead
svgPanZoom(
  svgPlot(
    show(m )
    # will have to manually size the svg device
    , height = 10, width = 16 )
)

```

How about using [`ggfortify`](https://github.com/sinhrks/ggfortify)?  Here are some great [examples](http://rpubs.com/sinhrks).  Let's make some pan and zoom.

```
# devtools::install_github('sinhrks/ggfortify')
library(ggfortify)
svgPanZoom(
  autoplot(lm(Petal.Width ~ Petal.Length, data = iris))
)

library(survival)
svgPanZoom(
  autoplot(survfit(Surv(time, status) ~ sex, data = lung))
)
```

If ternary diagrams excite you, let's do this [USDA soil example](http://www.ggtern.com/2014/01/15/usda-textural-soil-classification/) from [`ggtern`](http://www.ggtern.com).

```
install.packages("ggtern")
library(ggtern)

# Load the required libraries
library(ggtern)
library(plyr)
library(grid)
 
# Load the Data. (Available in ggtern 1.0.3.0 next version)
data(USDA)
 
# Put tile labels at the midpoint of each tile.
USDA.LAB = ddply(USDA, 'Label', function(df) {
    apply(df[, 1:3], 2, mean)
})
 
# Tweak
USDA.LAB$Angle = 0
USDA.LAB$Angle[which(USDA.LAB$Label == 'Loamy Sand')] = -35

# Construct the plot.
gTern <- ggplot(data = USDA, aes(y=Clay, x=Sand, z=Silt,
                        color = Label,
                        fill = Label)) +
  coord_tern(L="x",T="y",R="z") +
  geom_polygon(alpha = 0.75, size = 0.5, color = 'black') +
  geom_text(data = USDA.LAB,
            aes(label = Label, angle = Angle),
            color = 'black',
            size = 3.5) +
  theme_rgbw() +
  theme_showsecondary() +
  theme_showarrows() +
  custom_percent("Percent") +
  theme(axis.tern.padding    = unit(0.15, 'npc')) +
  labs(title = 'USDA Textural Classification Chart',
       fill  = 'Textural Class',
       color = 'Textural Class')
       
svgPanZoom(gTern)
```


The true test for me though will be financial time series plots.  Will `svgPanZoom` pass the test?

```
library(svgPanZoom)
library(SVGAnnotation)
library(PerformanceAnalytics)

data(edhec)

svgPanZoom(
  svgPlot(
    charts.PerformanceSummary(
      edhec
      ,main = "Performance of EDHEC Indicies"
    )
  )
)

library(quantmod)
getSymbols("JPM", from = "2013-12-31")
svgPanZoom(
  svgPlot(
    chartSeries(JPM, theme = chartTheme('white'),
        multi.col=T,TA="addVo();addBBands();addCCI()")
    ,height = 7
    ,width = 12
  )
)
```


Tal Galili's [`dendextend`](https://github.com/talgalili/dendextend) offers another great use case for pan and zoom interaction.  Let's look at one of the examples from the package vignette.

```
# install.packages("dendextend")
library(dendextend)
library(SVGAnnotation)
library(svgPanZoom)

data(iris) 
d_iris <- dist(iris[,-5]) # method="man" # is a bit better
hc_iris <- hclust(d_iris)
dend_iris <- as.dendrogram(hc_iris)
iris_species <- rev(levels(iris[,5]))
dend_iris <- color_branches(dend_iris,k=3, groupLabels=iris_species)
# have the labels match the real classification of the flowers:
labels_colors(dend_iris) <-
   rainbow_hcl(3)[sort_levels_values(
      as.numeric(iris[,5])[order.dendrogram(dend_iris)]
   )]

# We'll add the flower type
labels(dend_iris) <- paste(as.character(iris[,5])[order.dendrogram(dend_iris)],
                           "(",labels(dend_iris),")", 
                           sep = "")

dend_iris <- hang.dendrogram(dend_iris,hang_height=0.1)

# reduce the size of the labels:
dend_iris <- assign_values_to_leaves_nodePar(dend_iris, 0.5, "lab.cex")

par(mar = c(3,3,3,7))
svgPlot(
  {
    plot(dend_iris, 
         main = "Clustered Iris dataset
         (the labels give the true flower species)", 
         horiz =  TRUE,  nodePar = list(cex = .007))
    legend("topleft", legend = iris_species, fill = rainbow_hcl(3))
  }
  , height = 12, width = 14
) %>% svgPanZoom
```

For what I consider the ultimate test, will it work with `HiveR`?

```
# from HiveR documentation ?plotHive 
library(HiveR)
library(grid)
library(SVGAnnotation)
library(svgPanZoom)

data(HEC)
svgPanZoom(
  svgPlot(
    {
      data(HEC)
      currDir = getwd()
      setwd(system.file("extdata", "Misc", package = "HiveR"))
      plotHive(HEC, ch = 0.1, bkgnd = "white",
               axLabs = c("hair\ncolor", "eye\ncolor"),
               axLab.pos = c(1, 1),
               axLab.gpar = gpar(fontsize = 14),
               anNodes = "HECnodes.txt",
               anNode.gpar = gpar(col = "black"),
               grInfo = "HECgraphics.txt",
               arrow = c("more\ncommon", 0.0, 2, 4, 1, -2))
      
      grid.text("males", x = 0, y = 2.3, default.units = "native")
      grid.text("females", x = 0, y = -2.3, default.units = "native")
      grid.text("Pairing of Eye Color with Hair Color", x = 0, y = 3.75,
                default.units = "native", gp = gpar(fontsize = 18))
      grid.text("A test of plotHive annotation options", x = 0, y = 3.25,
                default.units = "native", gp = gpar(fontsize = 12))
      grid.text("Images from Wikipedia Commons", x = 0, y = -3.5,
                default.units = "native", gp = gpar(fontsize = 9))
      setwd(currDir)
    }
  , height = 20
  , width = 30)
)
```


NEWS
svgPanZoom 0.3
------------------------------------------------------------------------

* Upgrade to svg-pan-zoom.js v3.2.3
* Adds touch support with hammer.js
* Internally add better sizing to fit the htmlwidget container
* Add the ability to incorporate tasks on the JavaScript side

svgPanZoom 0.2
------------------------------------------------------------------------

* Initial release to CRAN


inst/htmlwidgets/svgPanZoom.js
HTMLWidgets.widget({

  name: 'svgPanZoom',

  type: 'output',

  initialize: function(el, width, height) {

    return {
      // TODO: add instance fields as required
    }

  },

  renderValue: function(el, x, instance) {
    // set innerHTML equal to the SVG provided by R as x.svg
    //  for better robustness, create DOM element separately
    //  and add to the htmlwidgets container el
    el.innerHTML = x.svg;

    var svg = el.getElementsByTagName("svg")[0]

    // use this to sort of make our diagram responsive
    //  or at a minimum fit within the bounds set by htmlwidgets
    //  for the parent container
    function makeResponsive(el){
       var svg = el.getElementsByTagName("svg")[0];
       if(svg){
        if(svg.width) {svg.removeAttribute("width")};
        if(svg.height) {svg.removeAttribute("height")};
        svg.style.width = "100%";
        svg.style.height = "100%";
       }
    };

    makeResponsive(el);

    // add touch with hammer.js
    //  using code from example
    //  https://github.com/ariutta/svg-pan-zoom/blob/master/demo/mobile.html
    x.config.customEventsHandler = {
      haltEventListeners: ['touchstart', 'touchend', 'touchmove', 'touchleave', 'touchcancel']
    , init: function(options) {
        var instance = options.instance
          , initialScale = 1
          , pannedX = 0
          , pannedY = 0
        // Init Hammer
        // Listen only for pointer and touch events
        this.hammer = Hammer(options.svgElement, {
          inputClass: Hammer.SUPPORT_POINTER_EVENTS ? Hammer.PointerEventInput : Hammer.TouchInput
        })
        // Enable pinch
        this.hammer.get('pinch').set({enable: true})
        // Handle double tap
        this.hammer.on('doubletap', function(ev){
          instance.zoomIn()
        })
        // Handle pan
        this.hammer.on('panstart panmove', function(ev){
          // On pan start reset panned variables
          if (ev.type === 'panstart') {
            pannedX = 0
            pannedY = 0
          }
          // Pan only the difference
          instance.panBy({x: ev.deltaX - pannedX, y: ev.deltaY - pannedY})
          pannedX = ev.deltaX
          pannedY = ev.deltaY
        })
        // Handle pinch
        this.hammer.on('pinchstart pinchmove', function(ev){
          // On pinch start remember initial zoom
          if (ev.type === 'pinchstart') {
            initialScale = instance.getZoom()
            instance.zoom(initialScale * ev.scale)
          }
          instance.zoom(initialScale * ev.scale)
        })
        // Prevent moving the page on some devices when panning over SVG
        options.svgElement.addEventListener('touchmove', function(e){ e.preventDefault(); });
      }
    , destroy: function(){
        this.hammer.destroy()
      }
    }

    instance.zoomWidget = svgPanZoom(svg, x.config);

    // add back viewBox that svgPanZoom removes to fill the container
    svg.setAttribute(
      'viewBox',
      ['0','0',svg.getClientRects()[0].width,svg.getClientRects()[0].height].join(' ')
    )

    // set up a container for tasks to perform after completion
    //  one example would be add callbacks for event handling
    //  styling
    if (!(typeof x.tasks === "undefined") ){
      if ( (typeof x.tasks.length === "undefined") ||
       (typeof x.tasks === "function" ) ) {
         // handle a function not enclosed in array
         // should be able to remove once using jsonlite
         x.tasks = [x.tasks];
      }
      x.tasks.map(function(t){
        // for each tasks call the task with el supplied as `this`
        t.call({el:el,zoomWidget:instance.zoomWidget});
      });
    }

    //  use expando property so we can access later
    //    somewhere saw where expando can cause memory leak in IE
    //    could also set in HTMLWidgets.widgets[x] where matches el
    el.zoomWidget = instance.zoomWidget;

  },

  resize: function(el, width, height, instance) {

  }

});


R/svgPanZoom.R
#' Pan and Zoom R graphics
#'
#' Add panning and zooming to almost any R graphics from base graphics,
#'    lattice, and ggplot2 by using the JavaScript library
#'     \href{https://github.com/ariutta/svg-pan-zoom}{svg-pan-zoom}.
#'
#' @param svg one of
#' \itemize{
#'   \item svg - SVG as XML, such as return from \code{\link[SVGAnnotation]{svgPlot}}
#'   \item lattice plot - trellis object, such as \code{l} in \code{l=xyplot(...)}
#'   \item ggplot2 plot - ggplot object, such as \code{g} in \code{g=ggplot(...) + geom_line()}
#'   \item filename or connection of a SVG file
#' }
#' @param ... other configuration options for svg-pan-zoom.js.
#'           See \href{https://github.com/ariutta/svg-pan-zoom#how-to-use}{svg-pan-zoom How To Use}
#'           for a full description of the options available.  As an example to turn on
#'           \code{controlIconsEnabled} and turn ,
#'           do \code{svgPanZoom( ..., controlIconsEnabled = TRUE, panEnabled = FALSE )}.
#' @param width,height valid CSS unit (like "100\%", "400px", "auto") or a number,
#'           which will be coerced to a string and have "px" appended
#' @param elementId \code{string} id for the \code{svgPanZoom} container.  Since \code{svgPanZoom}
#'           does not display its container, this is very unlikely to be anything other than the
#'           default \code{NULL}.
#'
#' @examples
#' #  svgPanZoom tries to be very flexible with its first argument
#'
#' #  in this first example use SVG as a character string
#' #    this is probably the least likely use case
#' library(svgPanZoom)
#' svgPanZoom('
#'    <svg style="height:300px;width:300px;">
#'      <circle cx="60" cy="60" r="50" style="fill:none;stroke:blue;"/>
#'    </svg>
#' ')
#'
#'
#' \dontrun{
#' library(svgPanZoom)
#'
#' # first let's demonstrate a base plot
#' # use svgPlot for now
#' library(SVGAnnotation)
#' svgPanZoom( svgPlot( plot(1:10) ) )
#'
#' svgPanZoom( svgPlot(show( xyplot( y~x, data.frame(x=1:10,y=1:10) ) ) )
#'
#' # the package gridSVG is highly recommended for lattice and ggplot2
#' # second let's demonstrate a lattice plot
#' library(lattice)
#' svgPanZoom( xyplot( y~x, data.frame(x=1:10,y=1:10) ) )
#'
#' # third with a ggplot2 plot
#' library(ggplot2)
#' svgPanZoom( ggplot( data.frame(x=1:10,y=1:10), aes(x=x,y=y) ) + geom_line() )
#'
#' #Of course as a good htmlwidget should, it works with Shiny also.
#' library(shiny)
#' library(SVGAnnotation)
#' library(svgPanZoom)
#' library(ggplot2)
#'
#' ui <- shinyUI(bootstrapPage(
#'   svgPanZoomOutput(outputId = "main_plot")
#' ))
#'
#' server = shinyServer(function(input, output) {
#'   output$main_plot <- renderSvgPanZoom({
#'     p <- ggplot() +
#'      geom_point(
#'        data=data.frame(faithful),aes(x=eruptions,y=waiting)
#'      ) +
#'      stat_density2d(
#'        data=data.frame(faithful)
#'        ,aes(x=eruptions,y=waiting ,alpha =..level..)
#'        ,geom="polygon") +
#'      scale_alpha_continuous(range=c(0.05,0.2))
#'
#'      svgPanZoom(p, controlIconsEnabled = T)
#'   })
#' })
#'
#' runApp(list(ui=ui,server=server))
#' }
#'
#' @export
svgPanZoom <- function(svg, ... , width = NULL, height = NULL, elementId = NULL) {

  # check to see if trellis for lattice or ggplot
  if(inherits(svg,c("trellis","ggplot","ggmultiplot"))){
    # if class is trellis then plot then use grid.export
    # try to use gridSVG if available
    if (requireNamespace("gridSVG", quietly = TRUE)) {
      print(svg)
      svg = gridSVG::grid.export(name=NULL)$svg
    } else {  #use SVGAnnotation
      if(requireNamespace("SVGAnnotation", quietly = TRUE)){
        warning("for best results with ggplot2 and lattice, please install gridSVG")
        svg = SVGAnnotation::svgPlot(svg, addInfo = F)
      } else { # if
        stop(
          "SVGAnnotation or gridSVG required with lattice or trellis objects",
           call. = FALSE
        )
      }
    }
  }

  # check to see if svg is XML and saveXML if so
  if(inherits(svg,c("XMLAbstractDocument","XMLAbstractNode"))){
    # should we add check for svg element?
    svg = XML::saveXML(svg)
  }

  # use SVG file if provided
  # thanks @jjallaire for code from https://github.com/rich-iannone/DiagrammeR
  # to use file or connection
  if ( inherits(svg, "connection") || ( class(svg) == "character" && file.exists(svg) ) ){
    # might want to parse to insure validity
    svg <- readLines(svg, warn = FALSE)
  }


  # forward options using x
  x = list(
    svg = svg
    ,config = list(...)
  )

  # create widget
  htmlwidgets::createWidget(
    name = 'svgPanZoom',
    x,
    width = width,
    height = height,
    package = 'svgPanZoom',
    elementId = elementId
  )
}

#' Shiny bindings for svgPanZoom
#'
#' @param outputId output variable to read from
#' @param width,height must be a valid CSS unit (like "100%", "400px", "auto") or a number,
#'           which will be coerced to a string and have "px" appended
#' @param expr \code{expression} that generates a svgPanZoom htmlwidget
#' @param env \code{environment} in which to evaluate \code{expr}
#' @param quoted \code{logical} is \code{expr} a quoted \code{expression} (with \code{quote()})?
#'           This is useful if you want to save an \code{expression} in a variable.
#'
#' @name svgPanZoom-shiny
#'
#' @export
svgPanZoomOutput <- function(outputId, width = '100%', height = '400px'){
  htmlwidgets::shinyWidgetOutput(outputId, 'svgPanZoom', width, height, package = 'svgPanZoom')
}

#' @rdname svgPanZoom-shiny
#'
#' @export
renderSvgPanZoom <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, svgPanZoomOutput, env, quoted = TRUE)
}
